xenomint/chain.go

```diff
@@ -0,0 +1,265 @@
+		cancel context.CancelFunc
+	)
+	// TODO(leventeliu): add multiple storage engine support.
+	if strg, err = xs.NewSqlite(filename); err != nil {
+		return
+	}
+	if state, err = newState(strg); err != nil {
+		return
+	}
+	if priv, err = kms.GetLocalPrivateKey(); err != nil {
+		return
+	}
+	ctx, cancel = context.WithCancel(context.Background())
+	c = &Chain{
+		state:  state,
+		in:     make(chan command, inCommandBufferLength),
+		out:    make(chan command, outCommandBufferLength),
+		priv:   priv,
+		wg:     &sync.WaitGroup{},
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	return
+}
+
+// Query queries req from local chain state and returns the query results in resp.
+func (c *Chain) Query(req *wt.Request) (resp *wt.Response, err error) {
+	var ref *query
+	if ref, resp, err = c.state.Query(req); err != nil {
+		return
+	}
+	if err = resp.Sign(c.priv); err != nil {
+		return
+	}
+	ref.updateResp(resp)
+	startWorker(c.ctx, c.wg, func(_ context.Context) {
+		c.enqueueOut(&applyRequest{request: req, response: resp})
+	})
+	return
+}
+
+func startWorker(ctx context.Context, wg *sync.WaitGroup, f func(context.Context)) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		f(ctx)
+	}()
+}
+
+func (c *Chain) stopWorkers() {
+	c.cancel()
+	c.wg.Wait()
+}
+
+func (c *Chain) enqueueIn(cmd command)  { c.in <- cmd }
+func (c *Chain) enqueueOut(cmd command) { c.out <- cmd }
+
+func (c *Chain) advanceTurn(ctx context.Context, t time.Time) {
+	defer func() {
+		c.stMu.Lock()
+		defer c.stMu.Unlock()
+		c.nextTurn++
+	}()
+	if !c.isMyTurn() {
+		return
+	}
+	// Send commit request to leader peer
+	var tCtx, cancel = context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+	var err error
+	if err = c.caller.CallNodeWithContext(tCtx, c.peers.Leader.ID, "", nil, nil); err != nil {
+		// ...
+	}
+}
+
+func (c *Chain) main(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// Try to fetch block of current turn
+			if !c.isMyTurn() {
+				// ...
+			}
+			//
+			if t, d := c.nextTick(); d > 0 {
+				time.Sleep(d)
+			} else {
+				// End current turn
+				c.advanceTurn(c.ctx, t)
+			}
+		}
+	}
+	return
+}
+
+func (c *Chain) processIn(ctx context.Context) {
+	var err error
+	for {
+		select {
+		case ci := <-c.in:
+			switch cmd := ci.(type) {
+			case *MuxApplyRequest:
+				if err = c.state.Replay(cmd.Request, cmd.Response); err != nil {
+					log.WithFields(log.Fields{
+						"request":  cmd.Request,
+						"response": cmd.Response,
+					}).WithError(err).Errorf("Failed to apply request")
+				}
+			default:
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *Chain) commitBlock() (err error) {
+	if err = c.state.commit(c.out); err != nil {
+		return
+	}
+	return
+}
+
+func (c *Chain) processOut(ctx context.Context) {
+	for {
+		select {
+		case ci := <-c.out:
+			switch ci.(type) {
+			case *applyRequest:
+			default:
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *Chain) Stop() (err error) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method Chain.Stop should have comment or be unexported